sdk/inc/azure/iot/az_iot_pnp_client.h

```diff
@@ -0,0 +1,706 @@
+ * @param[out] out_mqtt_user_name_length __[nullable]__ Contains the string length, in bytes, of \p
+ * mqtt_user_name. Can be `NULL`.
+ *
+ * @pre \p client must not be `NULL`.
+ * @pre \p mqtt_user_name must not be `NULL`.
+ * @pre \p mqtt_user_name_size must be greater than 0.
+ *
+ * @return An #az_result value indicating the result of the operation.
+ */
+AZ_NODISCARD az_result az_iot_pnp_client_get_user_name(
+    az_iot_pnp_client const* client,
+    char* mqtt_user_name,
+    size_t mqtt_user_name_size,
+    size_t* out_mqtt_user_name_length);
+
+/**
+ * @brief Gets the MQTT client ID.
+ *
+ * The client ID will be of the following format:
+ *
+ * **Without module ID**
+ *
+ * `{device_id}`
+ *
+ * **With module ID**
+ *
+ * `{device_id}/{module_id}`
+ *
+ * @param[in] client The #az_iot_pnp_client to use for this call.
+ * @param[out] mqtt_client_id A buffer with sufficient capacity to hold the MQTT client ID. If
+ * successful, contains a null-terminated string with the client ID that needs to be passed to the
+ * MQTT client.
+ * @param[in] mqtt_client_id_size The size, in bytes, of \p mqtt_client_id.
+ * @param[out] out_mqtt_client_id_length __[nullable]__ Contains the string length, in bytes, of \p
+ * mqtt_client_id. Can be `NULL`.
+ *
+ * @pre \p client must not be `NULL`.
+ * @pre \p mqtt_client_id must not be `NULL`.
+ * @pre \p mqtt_client_id_size must be greater than 0.
+ *
+ * @return An #az_result value indicating the result of the operation.
+ */
+AZ_NODISCARD az_result az_iot_pnp_client_get_client_id(
+    az_iot_pnp_client const* client,
+    char* mqtt_client_id,
+    size_t mqtt_client_id_size,
+    size_t* out_mqtt_client_id_length);
+
+/*
+ *
+ * SAS Token APIs
+ *
+ *   Use the following APIs when the Shared Access Key is available to the application or stored
+ *   within a Hardware Security Module. The APIs are not necessary if X509 Client Certificate
+ *   Authentication is used.
+ */
+
+/**
+ * @brief Gets the Shared Access clear-text signature.
+ *
+ * The application must obtain a valid clear-text signature using this API, sign it using
+ * HMAC-SHA256 using the Shared Access Key as password then Base64 encode the result.
+ *
+ * Use the SAS APIs when the Shared Access Key is available to the application or stored
+ * within a Hardware Security Module. The APIs are not necessary if X509 Client Certificate
+ * Authentication is used.
+ *
+ * @param[in] client The #az_iot_pnp_client to use for this call.
+ * @param[in] token_expiration_epoch_time The time, in seconds, from 1/1/1970.
+ * @param[in] signature An empty #az_span with sufficient capacity to hold the SAS signature.
+ * @param[out] out_signature The output #az_span containing the SAS signature.
+ *
+ * @pre \p client must not be `NULL`.
+ * @pre \p token_expiration_epoch_time must be greater than 0.
+ * @pre \p signature must be a valid, non-empty #az_span.
+ * @pre \p out_signature must not be `NULL`. It must point to an #az_span instance.
```

Comment:
nit: How can it point to anything but an `#az_span` instance? The compiler will catch this. Remove.
```suggestion
 * @pre \p out_signature must not be `NULL`.
```

Reply:
Original thought was conveying to users that it should be an address to an allocated `az_span`. Not something like this:
```c
az_span* span_ptr;
this_api_which_i_can't_see_the_name_of(&client, span_ptr);
```

Reply:
Wouldn't that, in many cases, make `span_ptr` be `NULL`, anyway?

Reply:
not if it's on the stack